tests/components/backup/common.py

```diff
@@ -161,57 +98,56 @@ async def get_backup(backup_id: str, **kwargs: Any) -> AgentBackup | None:
         return_value=backups, spec_set=[BackupAgent.async_list_backups]
     )
     mock_agent.async_upload_backup = AsyncMock(
-        spec_set=[BackupAgent.async_upload_backup]
+        side_effect=upload_backup,
+        spec_set=[BackupAgent.async_upload_backup],
     )
     return mock_agent
 
 
 async def setup_backup_integration(
     hass: HomeAssistant,
     with_hassio: bool = False,
-    configuration: ConfigType | None = None,
     *,
     backups: dict[str, list[AgentBackup]] | None = None,
     remote_agents: list[str] | None = None,
-) -> bool:
+) -> dict[str, Mock]:
     """Set up the Backup integration."""
+    backups = backups or {}
     with (
         patch("homeassistant.components.backup.is_hassio", return_value=with_hassio),
         patch(
             "homeassistant.components.backup.backup.is_hassio", return_value=with_hassio
         ),
     ):
         remote_agents = remote_agents or []
+        remote_agents_dict = {}
+        for agent in remote_agents:
+            if not agent.startswith(f"{TEST_DOMAIN}."):
+                raise ValueError(f"Invalid agent_id: {agent}")
+            name = agent.partition(".")[2]
+            remote_agents_dict[agent] = mock_backup_agent(name, backups.get(agent))
         platform = Mock(
             async_get_backup_agents=AsyncMock(
-                return_value=[BackupAgentTest(agent, []) for agent in remote_agents]
+                return_value=list(remote_agents_dict.values())
             ),
             spec_set=BackupAgentPlatformProtocol,
         )
 
         mock_platform(hass, f"{TEST_DOMAIN}.backup", platform or MockPlatform())
         assert await async_setup_component(hass, TEST_DOMAIN, {})
-
-        result = await async_setup_component(hass, DOMAIN, configuration or {})
+        assert await async_setup_component(hass, DOMAIN, {})
         await hass.async_block_till_done()
-        if not backups:
-            return result
 
-        for agent_id, agent_backups in backups.items():
-            if with_hassio and agent_id == LOCAL_AGENT_ID:
-                continue
-            agent = hass.data[DATA_MANAGER].backup_agents[agent_id]
+        if LOCAL_AGENT_ID not in backups or with_hassio:
+            return remote_agents_dict
 
-            async def open_stream() -> AsyncIterator[bytes]:
-                """Open a stream."""
-                return aiter_from_iter((b"backup data",))
+        agent = hass.data[DATA_MANAGER].backup_agents[LOCAL_AGENT_ID]
 
-            for backup in agent_backups:
-                await agent.async_upload_backup(open_stream=open_stream, backup=backup)
-            if agent_id == LOCAL_AGENT_ID:
-                agent._loaded_backups = True
+        for backup in backups[LOCAL_AGENT_ID]:
+            await agent.async_upload_backup(open_stream=None, backup=backup)
+        agent._loaded_backups = True
```

Comment:
Side note: This is still annoying. I guess the next step is to have real backup temp files that we can create via pytest fixtures and have the local agent just read from that space during setup.